File: DynamicKey/AgoraDynamicKey/nodejs/src/DynamicKey3.js
/**
* Server side. Generate Client key.
*/

var crypto = require('crypto');


module.exports.generate = function(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs) {
    channelName=channelName.toString();
    var version = "003";
    var unixTsStr = unixTs.toString();  //Unix Time stamp, track time as a running total of seconds
    var rndTxt = randomInt.toString(16);
    var randomIntStr = ("00000000" + rndTxt).substring(rndTxt.length);
    var uidStr = ("0000000000" + uid).substring(String(uid).length);
    var expiredTsStr = ("0000000000" + expiredTs).substring(String(expiredTs).length);
    var sign = generateSignature3(appID, appCertificate, channelName, unixTsStr, randomIntStr, uidStr, expiredTsStr);
    return version + sign + appID + unixTsStr + randomIntStr+ uidStr+ expiredTsStr;
};

var generateSignature3 = function(appID, appCertificate, channelName, unixTsStr, randomIntStr,uidStr, expiredTsStr ) {
    var buffer = Buffer.concat([new Buffer(appID), new Buffer(unixTsStr), new Buffer(randomIntStr), new Buffer(channelName), new Buffer(uidStr), new Buffer(expiredTsStr)]);
    var sign = encodeHMac(appCertificate, buffer);
    return sign.toString('hex');
};


var encodeHMac = function(key, message) {
    return crypto.createHmac('sha1', key).update(message).digest('hex');
};

File: DynamicKey/AgoraDynamicKey/go/src/DynamicKey3/DynamicKey3.go
package DynamicKey3

import (
	"crypto/hmac"
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"strings"
)

func Generate(appID, appCertificate, channelName string, unixTs, randomInt, uid, expiredTs uint32) (string) {
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs)
}

func generateDynamicKey(appID, appCertificate, channelName string, unixTs, randomInt, uid, expiredTs uint32) (string) {
    version := "003"
	unixTsStr := fmt.Sprintf("%010d", unixTs)
	randomIntStr := fmt.Sprintf("%08x", randomInt)
	uidStr := fmt.Sprintf("%010d", uid)
	expiredTsStr := fmt.Sprintf("%010d", expiredTs)
	signature := generateSignature(appID, appCertificate, channelName, unixTsStr, randomIntStr, uidStr, expiredTsStr)
	buffer := strings.Join([]string{version, signature, appID, unixTsStr, randomIntStr, uidStr, expiredTsStr}, "")
	return buffer
}

func generateSignature(appID, appCertificate, channelName, unixTsStr, randomIntStr, uidStr, expiredTsStr string) (string) {
	buffer := strings.Join([]string{appID, unixTsStr, randomIntStr, channelName, uidStr, expiredTsStr}, "")
	signature := hmac.New(sha1.New, []byte(appCertificate))
	signature.Write([]byte(buffer))
	return hex.EncodeToString(signature.Sum(nil))
}
